### ISSUE TITLE
Git Flow

### ISSUE THREAD

**Issue by ditchfieldcaleb** [fei-protocol/fei-protocol-core#594]
The Problem: Our git flow is not very analogous to what actually happens, especially around dao proposals, and often causes 3-4 day periods where we expect integration tests to fail. Also, we often merge code to our develop/working branch that never hits mainnet.

Let's discuss potential git flows that are more suited for our use-case.

Statements that seem applicable/helpful to this situation, mostly subjective:
- we should have a branch that represents "the canonical, current state of the protocol"
- we should have a branch that represents "the expected near-future state of the protocol" so that developers can rebase off of others' work for parallel, but dependent features
- we should not merge code to the branch that represents "production" until it actually hits production (ie, the dao transaction executes)
- the branch that represents "production" should almost never have failing tests - unit or integration

Right now our git flow involves merging code to develop when it is complete & ready to go to the dao. Thanks to some awesome work by @thomas-waite , the relevant integration tests fail less often, and dao proposals get simulated when needed more often, but this hasn't solved everything. We have a theoretical "release" process, but no metric for when to make a release or what master represents other than "more absolute than develop".

**Comment by ditchfieldcaleb** [comment]
Okay, so the below is long, and dense, but please give it your full attention if you can; I've put a lot of thought into our git flow and would appreciate it if you did too, it's an important issue to solve!

Current Git Flow:
```
develop--------------------+------------------->
       \--------feature------/ 
```

Proposed New Git Flow:

```
develop------------------------------------+----------------->
       \--------feature------\                       /
staging---------------------\-------------/------------------->
```

I think this isn't too complicated; here's the flow.

Happy path: Developer works on feature by branching off of develop. Developer finishes code, and merges into staging while the DAO vote is pending. Should the DAO vote succeed and the changes execute on mainnet, staging merges into develop and his changes are considered canonical.

Sad path: Developer works on feature by branching off of develop. Developer finishes code, and merges into staging while the DAO vote is pending. Should the DAO vote fail, or a technical error occur, and the changes _not_ execute on mainnet, the developer will _revert_ the changes on staging.

Sitenote: for changes that don't go to the dao, or execute immediately, there's no need to push these to staging, and we can merge these straight to the develop branch as usual.

The above approach uses a single staging branch so that developers working on features can rebase off of it for shared base code. But we have an issue: when we merge staging into develop, more than likely we'll be merging some code that hasn't been executed yet. What do?

Simple - instead of actually merging the entire branch, we cherry-pick the commits for the code-that-needs-to-be-merged and merge _only these_ into develop. This sounds complicated but is vastly simplified if we ensure that developers merging into staging squash down to a single commit first, which is good practice anyways.

Because staging is branched off of develop, as soon as we merge this cherry-picked commit into develop, it's no longer in staging! (because the diff is empty). This solves both the problem of "working off of a branch that only contains production code" and the problem of "needing to work off of pending-but-not-yet-executed" code simultaneously.

**Comment by thomas-waite** [comment]
Like the idea of refactoring our Git Flow a lot. Agreed that it's not great that we have a period where we expect e2e tests to be failing + has always felt a bit weird that we don't have a "release" framework or cycle and our codebase doesn't necessarily reflect what is on-chain. 

Some questions/thoughts:
1. Why when we merge `staging` into `develop` would: `more than likely we'll be merging some code that hasn't been executed yet. What do?` As I understand it, code should only go into develop once it has been fully executed on-chain?
2. Why not rename `develop` to `master` under this framework? It is effectively becoming a `master` branch
3. Is it worthwhile introducing some concept of 'tagging releases' as it is now clear when we are effectively making a release (i.e. when we merge to `develop`)? We could use `git tag` on the merge commit, to identify a version perhaps associated with the deployment/release

**Comment by eswak** [comment]
What you describe looks like a classic git flow setup where `develop` is what you call `staging` and where `master` is what you call `develop`, or did I miss something ? Seems like we just use a wrong terminology for our branches and lack some discipline to make the process match with our constraints (on-chain state & proposals process).

I don't think the proposed change will solve the issue we have : 
- Currently
  - stable unit & e2e tests on `master`
  - unit & e2e tests sometimes broken on `develop`
- Proposed (Caleb)
  - stable unit & e2e tests on `master` (and unclear what `master` has in addition to `develop`)
  - stable unit & e2e tests on `develop`
  - unit & e2e tests sometimes broken on `staging`

I think we're overthinking and can keep a normal git flow
```
master---------------------------+--------------------------------->
                                /
develop----\-----------+-------x----------------------------------->
            \         /
feature------\-+-+-o-/--------------------------------------------->
```

or, for more complicated features that are co-dependent, group them in an "epic" :
```
master---------------------------------------------+--------------->
                                                  /
develop-\-------------------------------------+--x----------------->
         \                                   /
epic------\-----------+-\---------------+-o-/---------------------->
           \         /   \             /
feature1----\-+-+-+-/     \           /
                           \         /
feature2--------------------\-+-+-+-/------------------------------>
```

This is identical to what we're doing now, with some small changes : 
- `o` commits make sure that `test/integration/proposals_config.ts` is always empty, because develop reflects the current state of mainnet without active proposals. We only merge a feature after it executed onchain.
- `x` commits are for major releases, and increment the version in `package.json` for instance.
- `+` commits are merge commits. I don't like squashing because history is lost, and cherry-picking affects a new commit hash to the same diff which is a bit of a brainfuck. Also some websites are counting commits & if we squash we will have rookie numbers, we gotta pump those numbers up 😅 

TBH I don't think we need to separate `master` and `develop` for our use case, we could just have a `main` branch that represents mainnet & everything else is just atomic features / epics that are waiting for on-chain execution. If something does not pass on mainnet, we don't merge the feature/epic branch, and case closed.

The final ascii art would look like this : 
```
feature2---------/-+-+-o-\----------------------------------------->
                /         \
main----\------/-----------+------------------+--x----------------->
         \                                   /
epic1-----\-----------+-\---------------+-o-/---------------------->
           \         /   \             /
feature1.1--\-+-+-+-/     \           /
                           \         /
feature1.2------------------\-+-+-+-/------------------------------>
```

**Comment by Joeysantoro** [comment]
I tend to agree with @eswak but perhaps keeping develop is useful. Ideally features can get merged to develop with a "dirty" proposals config so interactions between live proposals can be observed.

Our main issue honestly is discipline, and that may be because we haven't fully agreed on a git flow we're all ready to own. 

I'd like more "why" and what this does for development. The integration tests seem to be the main issue, so perhaps we just move our monitoring to the master branch instead of develop and enforce that master never has any live proposals. i.e. we do releases every week that includes a DAO vote.

**Comment by ditchfieldcaleb** [comment]
@Joeysantoro: Releases every week (when code with dao vote needs to be merged) seem decent to me.
@eswak: I like your counter-proposal but I think it requires more coordination and discipline than the current system; keeping track of the epics specfically seems like it could get messy.
@thomas-waite, answers to your questions:

> Why when we merge staging into develop would: more than likely we'll be merging some code that hasn't been executed yet. What do? As I understand it, code should only go into develop once it has been fully executed on-chain?
In my proposal there exists a single staging branch, that would have multiple pending efforts in it at any given time, resolving linearly, so merging the whole branch into develop would merge both the currently-finalized-and-executed proposal *and* the other pending proposals in the branch.

> Why not rename develop to master under this framework? It is effectively becoming a master branch
Fine by me, mostly just semantics here

> Is it worthwhile introducing some concept of 'tagging releases' as it is now clear when we are effectively making a release (i.e. when we merge to develop)? We could use git tag on the merge commit, to identify a version perhaps associated with the deployment/release. When we effectively make a release by merging from staging to develop, I wonder if there's scope to do other release related activities. How many people are building with Fei contracts? We could add a release flow here that builds and publishes an npm package with the latest contracts. Would definitely be useful at hackathons for the least

Both very good points; we'd need to identify a good place to tag releases (milestone or time interval) and ensure we're all on board for this, and probably assign someone specifically to be in charge of this, however.

**Comment by ditchfieldcaleb** [comment]
After reviewing everyone's feedback I feel like the most pareto solution is to keep our current system, but move monitoring and alerting to master like @Joeysantoro suggested, and add in a regular weekly (or other time interval) release schedule that someone (this could be a rotating person) is in charge of executing.

**Comment by ElliotFriedman** [comment]
Reading through this thread, I'm in agreement that we should just keep our current git flow and add a more regular merge from develop to master. Once a month is often enough, and we could possibly have a 1 hour meeting to review all changes that made it through from dev to master in the past month so that everyone gets a higher level view of what is happening.

**Comment by ditchfieldcaleb** [comment]
The issue that was brought up with this current proposal Is that any regular cadence of merging from dev ---> master would bring in changes that shouldn't be brought to master yet. This could potentially always be an issue if there is a proposal in-flight continuously across any period - then merging to master during that period would mess stuff up.

**Comment by Joeysantoro** [comment]
We need to sandbox non-production ready code. Imo this is the "epic" concept in some form.

Develop being a mix of prod and random contracts that shouldn't be deployed is a recipe for disaster. IF we can commit to keeping develop free of non-prod code then I think regular release branches is totally fine. It should be a branch that is cleaned up (removing proposals config and such and waiting for the proposals to complete on-chain) rather than a straight merge from develop->master

**Comment by ditchfieldcaleb** [comment]
What if we kept the system as-is and didn't merge into develop until code is executed?

If there are multiple features that are dependent on some base code, we can use epic branches.

**Comment by ditchfieldcaleb** [comment]
Will write up what we decided in dev meeting shortly.

**Comment by ditchfieldcaleb** [comment]
Branches:
- develop: what we branch off of; does not matter much if tests pass
- feature/x: standard feature branches, branched from develop, and *merged into staging*
- staging/x: new branch every week. features that are complete and sent to the dao are merged to this.
- master: should reflect the state of mainnet; all tests should always pass

Process:
**To work on a feature that does *not* require a dao vote:**
- branch off of develop. code your feature. once complete and pr review passed, rebase off of develop to pull in any changes that happened while you were working, merge into master after ensuring all tests pass, and then *merge develop into master*

**To work on a feature that *does* require a dao vote:**
- branch off of develop. code your feature. once complete and pr review passed, merge into the current staging branch

**As the on-call developer, every Thursday**:
- clean up the previous staging branch, ensuring that proposals-config is clear and all tests pass. merge this to master. then merge master into develop.
- lock the current staging branch and allow no more prs to be merged into it. the following week, this will get merged to master.
- create a new staging branch off of master for developers to merge to. this will get merged to master 2 weeks from now.

FAQ:
- I want to work on a feature that requires code that's still in a staging branch.
You can do this by rebasing your feature branch on top of the feature branch it depends on. Importantly, *don't branch off of the base-code feature branch.*. Instead, branch off of develop first like normal, and *then* rebase off of the feature branch your code will depend on.

- I am confused / this is too much process / can we not do this?
No. This is probably the simplest way we can make this work and have high-signal tests. It may sound a little complicated at first, but run through it a few times and the flow should start to get familiar. Make sure you know how rebases work.

```

 ---develop---+---------------------------------------+------>
               \                                     /
                +---feature/noDAOVote---+           /    
                                         \         /     
                                          \       /   
                                           \     /  
                                            \   /      
                                             \ /    
 ---master------------------------------------+-------------->
 
 

 ---develop---+-------------------------------------------------------------------------------------------+--->
               \                                                                                         /
                +---feature/daoVote---+                                                                 /
                                       \                                                               /
                                        \                                                             /
                +---staging/weekx--------+---(lock @ 1 week, for 1 week)-+-(cleanup & fixes)---+     /
                                                                                                \   /
                                                                                                 \ /
 ---master----+-----------------------------------------------------------------------------------+----------->
```

**Comment by Joeysantoro** [comment]
> Branches:
> 
> * develop: what we branch off of; does not matter much if tests pass
> * feature/x: standard feature branches, branched from develop, and _merged into staging_
> * staging/x: new branch every week. features that are complete and sent to the dao are merged to this.
> * master: should reflect the state of mainnet; all tests should always pass
> 
> Process: **To work on a feature that does _not_ require a dao vote:**
> 
> * branch off of develop. code your feature. once complete and pr review passed, rebase off of develop to pull in any changes that happened while you were working, merge into master after ensuring all tests pass, and then _merge develop into master_
> 
> **To work on a feature that _does_ require a dao vote:**
> 
> * branch off of develop. code your feature. once complete and pr review passed, merge into the current staging branch
> 
> **As the on-call developer, every Thursday**:
> 
> * clean up the previous staging branch, ensuring that proposals-config is clear and all tests pass. merge this to master. then merge master into develop.
> * lock the current staging branch and allow no more prs to be merged into it. the following week, this will get merged to master.
> * create a new staging branch off of master for developers to merge to. this will get merged to master 2 weeks from now.
> 
> FAQ:
> 
> * I want to work on a feature that requires code that's still in a staging branch.
>   You can do this by rebasing your feature branch on top of the feature branch it depends on. Importantly, _don't branch off of the base-code feature branch._. Instead, branch off of develop first like normal, and _then_ rebase off of the feature branch your code will depend on.
> * I am confused / this is too much process / can we not do this?
>   No. This is probably the simplest way we can make this work and have high-signal tests. It may sound a little complicated at first, but run through it a few times and the flow should start to get familiar. Make sure you know how rebases work.
> 
> ```
> 
>  ---develop---+---------------------------------------+------>
>                \                                     /
>                 +---feature/noDAOVote---+           /    
>                                          \         /     
>                                           \       /   
>                                            \     /  
>                                             \   /      
>                                              \ /    
>  ---master------------------------------------+-------------->
>  
>  
> 
>  ---develop---+-------------------------------------------------------------------------------------------+--->
>                \                                                                                         /
>                 +---feature/daoVote---+                                                                 /
>                                        \                                                               /
>                                         \                                                             /
>                 +---staging/weekx--------+---(lock @ 1 week, for 1 week)-+-(cleanup & fixes)---+     /
>                                                                                                 \   /
>                                                                                                  \ /
>  ---master----+-----------------------------------------------------------------------------------+----------->
> ```

Seems about right, some notes:
* why merge features to staging, imo should be just develop and develop -> staging
* should add epic/x naming convention for batched features like audits
* release might be a better name than staging but don't feel strongly

**Comment by Joeysantoro** [comment]
```
                                                                                    +-----epic/auditX----+-----+(more features)
                +---feature/daoVote---+                         /
               /                                        \                        /                         
 ---develop---+-------------------------------------------------------------------------------------------+--->
                                  \                                                                                            \                                                         
                                   \                                                                                            +-----staging/weekx+1                                                                     
                       +---staging/weekx-+-(lock for 1 week)-+-(cleanup & fixes)---+          
                                                                                                                               \   
                                                                                                                                \ 
 ---master----+-----------------------------------------------------------------------------------+----------->
```

Summary:
* same DX for devs of merging to develop with the caveat that develop MUST be protected from "non-prod" ready features
* on-call dev merges prior staging branch AND is responsible for cleanup (ideally just wipe proposals_config) before starting new staging branch
* hotfixes to master are merged back to develop. Normal tooling features merge to develop and land in master following cycle

**Comment by ditchfieldcaleb** [comment]
Summary, Expanded:

*Branches*
- develop: where you branch off of for features & staging. must not have stuff that isn't prod-ready, but can have stuff that is pending dao execution
- master: where all tests must pass. staging is merged into master weekly. then master is merged back into develop (if needed)
- staging/week-n-1: previous week's staging. on-call dev does any cleanup necessary and merges this into master
- staging/week-n: current' weeks' staging. on-call dev creates this by branching off of develop and It is locked for one week.
- epic: collection of features dependent on some base code that isn't prod-ready. otherwise treated as a feature branch.

**Comment by ditchfieldcaleb** [comment]
Closing issue.
